Properly configure metro.config.js for Expo

Yonom opened the pull request:
Fixes #141 by updating the documentation for metro.config.js configuration in Expo environments

# Why?

Currently, the metro config is not properly configured and this causes the following error from expo:

```
It looks like that you are using a custom metro.config.js that does not extend @expo/metro-config.
This can result in unexpected and hard to debug issues, like missing assets in the production bundle.
We recommend you to abort, fix the metro.config.js, and try again.
```

# Test Plan

Update the configuration in a project
Run eas build and ensure that no warning is emitted

kristerkari commented:
Thanks @Yonom ! Did you have some example of this change running somewhere? I think that this one needs to be properly tested before merging.

tidusjar commented:
Seemed to be working for me!

Yonom commented:
We have been using this in production for about 2 months without issues

Yonom commented:
Expo does not always use `metro-react-native-babel-transformer` as its babel transformer. I have added support for custom upstream transformers and updated the README to forward expo's transformer. This adds support for expo projects using the "exotic" transformer.

https://github.com/expo/expo-cli/blob/master/packages/metro-config/src/ExpoMetroConfig.ts#L262

giautm commented:
Oh wow, this is why the new transformer - exotic don't work for me. I think I will use patch-package.

giautm commented:
I have some confusion, `babelTransformerPath` looks like a string, the path point to the transformer directory. So, why did this PR change it to a function?

giautm commented on README.md:
```diff
@@ -58,13 +58,20 @@ const { getDefaultConfig } = require("expo/metro-config");
-      babelTransformerPath: require.resolve("react-native-svg-transformer")
+      ...transformer,
+      babelTransformerPath: require
+        .resolve('react-native-svg-transformer')
+        .makeTransformer(babelTransformerPath),
```
It's not correct.
```
require.resolve(...).makeTransformer is not a function
TypeError: require.resolve(...).makeTransformer is not a function
```

Yonom replied:
Did you apply the changes to index.js? https://github.com/kristerkari/react-native-svg-transformer/pull/151/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346

giautm replied:
require.resolve() return a string, so you can't call the export functions. Yes, I have patch the module and it didn't work.

Finally, I end up with custom metro-svg-transformer.js with content from the orignal transformer.

giautm commented:
`metro.config.js`:
```js metro.config.js
    transformer: {
      ...defaultConfig.transformer,
      babelTransformerPath: require.resolve('./metro-svg-transformer'),
    },
```

`metro-svg-transformer.js`, I also upgraded to the latest version of `@svgr/core`.
```js metro-svg-transformer.js
const { getDefaultConfig } = require('@expo/metro-config')
const { transform: svgrTransform, resolveConfig } = require('@svgr/core')
const path = require('path')

const defaultSvgrConfig = {
  native: true,
  plugins: ['@svgr/plugin-svgo', '@svgr/plugin-jsx'],
  svgoConfig: {
    plugins: [
      {
        name: 'inlineStyles',
        active: true,
        params: {
          onlyMatchedOnce: false,
        },
      },
      {
        name: 'removeViewBox',
        active: false,
      },
      {
        name: 'removeUnknownsAndDefaults',
        active: false,
      },
      {
        name: 'convertColors',
        active: false,
      },
    ],
  },
}

function makeTransformer() {
  const config = getDefaultConfig(__dirname)
  const { babelTransformerPath } = config.transformer
  const transformer = require(babelTransformerPath)

  return function (src, filename, options) {
    if (typeof src === 'object') {
      // handle RN >= 0.46
      ;({ src, filename, options } = src)
    }

    if (filename.endsWith('.svg') || filename.endsWith('.svgx')) {
      const config = resolveConfig.sync(path.dirname(filename))
      const svgrConfig = config
        ? Object.assign({}, defaultSvgrConfig, config)
        : defaultSvgrConfig

      const jsCode = svgrTransform.sync(src, svgrConfig)
      return transformer.transform({
        src: jsCode,
        filename,
        options,
      })
    }

    return transformer.transform({ src, filename, options })
  }
}

exports.transform = makeTransformer()
```

mersiades commented:
For what it's worth, @Yonom 's solution gives me 

```
require.resolve(...).makeTransformer is not a function
TypeError: require.resolve(...).makeTransformer is not a function
```

as well. @giautm 's solution gets me a bit closer, but:
![IMG_122A9DAB2373-1](https://user-images.githubusercontent.com/6284016/148502559-e5dd7319-b957-4952-b125-204c8120bd9b.jpeg)

giautm commented:
Hey @mersiades, please show your config/ react-native-svg, svgr version.

I have been tested it with SDKv43, SDKv44 without any problem.

mersiades commented:
> Hey @mersiades, please show your config/ react-native-svg, svgr version.

I'm on SDKv43,  and svgr and react-native-svg are both at 1.22.4.

My immediate problem re #141 has been solved; just posting these errors here in case they're useful to the PR.

giautm commented:
> > Hey @mersiades, please show your config/ react-native-svg, svgr version.
> 
> 
> 
> I'm on SDKv43,  and svgr and react-native-svg are both at 1.22.4.
> 
> 
> 
> My immediate problem re #141 has been solved; just posting these errors here in case they're useful to the PR.

Please share the repo with me, So I can check the error. You also can DM to me via Telegram or Twitter.

mersiades commented:
> > My immediate problem re #141 has been solved; just posting these errors here in case they're useful to the PR.
> 
> Please share the repo with me, So I can check the error. You also can DM to me via Telegram or Twitter.

Sorry, it's a private repo. But if there's anything else I can do to help, please ask.

jameswilddev commented:
Hey!  Is there anything I can do to help move this along?  We're looking to migrate to EAS ASAP and we think this is the root cause of our build failures over there.

Yonom commented:
Closing this as I unfortunately did not find the time to fix the PR.
The problem with the current upstream implementation is that the expo transformer is replaced with metro's default. This is not always correct for expo projects (and causes things like the [exotic transformer](https://blog.expo.dev/drastically-faster-bundling-in-react-native-a54f268e0ed1) to break)
